interp/src/stk_env.rs

```diff
@@ -0,0 +1,809 @@
+    /// node onto the fork point. Returns the resulting Smoosher.
+    /// ```
+    /// ## IMPORTANT INVARIANT
+    /// ```text
+    /// The intersection of the bindings of the two branches
+    /// MUST be the empty set! Otherwise, no guarantee for which binding will be included
+    /// in the merge.
+    /// ```
+    /// # Panics
+    /// ```text
+    /// Panics if [self] and [other] do not share a common fork point, or if either is [empty]
+    /// ```
+    /// # Examples
+    /// ## Pictorial Example
+    /// ```text
+    /// Smoosher 1: (A, B, C, F, G) Smoosher 2: (D, E, F, G)          
+    /// [A]
+    ///  |
+    /// [B]        [D]
+    ///  |          |
+    /// [C]        [E]
+    ///  |          |
+    ///   \        /
+    ///    \      /
+    ///      [F]
+    ///       |
+    ///      [G]
+    /// *merge(Smoosher_1, Smoosher_2)
+    /// Returns:
+    ///      [ABCDE merged and smooshed onto F]
+    ///       |
+    ///      [G]
+    /// ```
+    /// ## Code Example
+    /// good:
+    /// ```
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// a.set("hi!", 1);
+    /// a.new_scope();
+    /// a.set("bye", 0);
+    /// let mut b = a.fork();
+    /// a.set("hi!", 3);
+    /// b.set("bye", 2);
+    /// let c = Smoosher::merge(a, b);
+    /// assert_eq!(*c.get(&"hi!").unwrap(), 3);
+    /// assert_eq!(*c.get(&"bye").unwrap(), 2);
+    /// assert_eq!(*c.get(&"hello").unwrap(), 15);
+    /// ```
+    /// will panic:
+    /// ```should_panic
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// let mut b = Smoosher::new();
+    /// b.set("hey", 13);
+    /// let c = Smoosher::merge(a, b); //there's no common fork point
+    /// ```
+    pub fn merge(self, other: Self) -> Self {
+        //find shared fork point; if doesn't exist, panic
+        let mut a = self;
+        let mut b = other;
+        if let Some((depth_a, depth_b)) = Smoosher::shared_fork_point(&a, &b) {
+            //smoosh [self] and [other] to right before that point
+            a = a.smoosh(depth_a - 1);
+            b = b.smoosh(depth_b - 1);
+
+            //now do merge
+
+            //get both heads of smooshed a and b
+            let mut a_head = a.head;
+            //merge a_head and b_head.
+            for (k, v) in b.head {
+                a_head.insert(k, v);
+            }
+            //now drop b
+            std::mem::drop(b.tail); //b.head already consumed above
+                                    //create a'
+            let (a_new_head, a_new_tail) = a.tail.split();
+            if let Some(mut a_new_head) = a_new_head {
+                a = Smoosher {
+                    head: a_new_head,
+                    tail: a_new_tail,
+                };
+            } else {
+                panic!("trying to merge, but [self] is empty")
+            }
```

Comment:
Why should this be a panic? Merging with an empty list is just the identity function, no?

Reply:
this shouldn't really be possible, because to merge two Smooshers they must share a common fork point. So they must both be at least 2 scopes large (though those scopes can be empty)

Reply:
so i will actually just change the text, the panic is in the right location IMO

Reply:
You can use the `unreachable!(...)` macro to denote that this code should be unreachable instead of panicking so that people in the future know that this is an invariant.